app/uk/gov/hmrc/cgtpropertydisposals/service/TaxEnrolmentService.scala

```diff
@@ -18,60 +18,61 @@ package uk.gov.hmrc.cgtpropertydisposals.service
-import uk.gov.hmrc.cgtpropertydisposals.service.TaxEnrolmentService.TaxEnrolmentResponse
-import uk.gov.hmrc.cgtpropertydisposals.service.TaxEnrolmentService.TaxEnrolmentResponse.TaxEnrolmentCreated
-import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.cgtpropertydisposals.models.{Error, TaxEnrolmentRequest}
+import uk.gov.hmrc.cgtpropertydisposals.repositories.TaxEnrolmentRepository
+import uk.gov.hmrc.cgtpropertydisposals.util.Logging
+import uk.gov.hmrc.http.{HeaderCarrier, HttpResponse}
 
 import scala.concurrent.ExecutionContext.Implicits.global
 import scala.concurrent.Future
 
 @ImplementedBy(classOf[TaxEnrolmentServiceImpl])
 trait TaxEnrolmentService {
-  def allocateEnrolmentToGroup(cgtReference: String, subscriptionDetails: SubscriptionDetails)(
+  def allocateEnrolmentToGroup(taxEnrolmentRequest: TaxEnrolmentRequest)(
     implicit hc: HeaderCarrier
-  ): EitherT[Future, Error, TaxEnrolmentResponse]
-}
-
-object TaxEnrolmentService {
-
-  sealed trait TaxEnrolmentResponse
-
-  object TaxEnrolmentResponse {
-    case object TaxEnrolmentCreated extends TaxEnrolmentResponse
-  }
-
+  ): EitherT[Future, Error, Unit]
 }
 
 @Singleton
-class TaxEnrolmentServiceImpl @Inject()(connector: TaxEnrolmentConnector) extends TaxEnrolmentService {
+class TaxEnrolmentServiceImpl @Inject()(
+  taxEnrolmentConnector: TaxEnrolmentConnector,
+  taxEnrolmentRetryRepository: TaxEnrolmentRepository
+) extends TaxEnrolmentService
+    with Logging {
 
-  override def allocateEnrolmentToGroup(cgtReference: String, subscriptionDetails: SubscriptionDetails)(
+//  @SuppressWarnings(Array("org.wartremover.warts.NonUnitStatements"))
+  override def allocateEnrolmentToGroup(taxEnrolmentRequest: TaxEnrolmentRequest)(
     implicit hc: HeaderCarrier
-  ): EitherT[Future, Error, TaxEnrolmentResponse] = {
+  ): EitherT[Future, Error, Unit] =
+    for {
+      httpResponse <- taxEnrolmentConnector
+                       .allocateEnrolmentToGroup(taxEnrolmentRequest)
+                       .leftMap(e => Error(s"Error calling tax enrolment service: $e"))
+      result <- EitherT.fromEither(handleTaxEnrolmentResponse(httpResponse)).leftFlatMap[Unit, Error] {
+                 (error: Error) =>
+                   logger
+                     .warn(s"Failed to allocate enrolments due to error: $error. Inserting enrolment details in mongo.")
+                   val dbResponse = for {
+                     writeResult <- taxEnrolmentRetryRepository
+                                     .insert(taxEnrolmentRequest)
+                                     .leftMap(error => Error(s"Error inserting enrolment details into mongo: $error"))
+                   } yield (writeResult)
+
+                   dbResponse
+                     .subflatMap { writeResult =>
+                       if (writeResult) Right(()) else Left(Error("Failed to insert enrolment details into mongo"))
+                     }
+                     .leftMap(error => error)
```

Comment:
you don't need this `leftMap` here since you're just doing an identity operation in it